spec/index.md

```diff
@@ -4400,19 +4400,30 @@ Env = {
 )
 or
 (
-  A.mode = upgrade {skip_pre_upgrade = true}
+  (A.mode = upgrade U and U.skip_pre_upgrade = true)
   Stable_memory = Old_State.stable_mem
   New_certified_data = NoCertifiedData
   Cycles_used = 0
 )
 
+(
+  (A.mode = upgrade U and U.wasm_memory_persistence ≠ keep)
+  Intermediate_state = {store = initial_wasm_store; self_id = A.canister_id; stable_mem = Stable_memory}
+)
+or
+(
+  (A.mode = upgrade U and U.wasm_memory_persistence = keep)
+  Intermediate_state = {store = load_data_segments(Old_state.store, A.wasm_module); self_id = A.canister_id; stable_mem = Stable_memory}
```

Comment:
For simplicity, we thought not to load the new active data segments and not to patch the memory (as this would need a reserved space and is also more complicated for the IC execution layer). Instead, we can use passive data segments that can be lazily loaded at a dynamic address.

It would be more like:
```suggestion
  Intermediate_state = {store = Old_state.store; self_id = A.canister_id; stable_mem = Stable_memory}
```